pywt/_extensions/_dwt.pyx

```diff
@@ -98,9 +98,9 @@ cpdef dwt_axis(np.ndarray data, Wavelet wavelet, MODE mode, unsigned int axis=0)
     return (cA, cD)
 
 
-cpdef idwt_single(data_t[::1] cA, data_t[::1] cD, Wavelet wavelet, MODE mode):
-    cdef data_t[::1] rec
+cpdef idwt_single(np.ndarray cA, np.ndarray cD, Wavelet wavelet, MODE mode):
```

Comment:
Also, I figure out why your example above succeeds.  `.astype('float')` and `.astype('double')` both give arrays with `dtype` as `np.float64` since the default `float` is 64-bit in python.  

If I modified the call as follows we see the expected failure:
```python
pywt._extensions._dwt.idwt_single(cA.astype(np.float32), cD.astype('double'), w, m)
```
```
  File "pywt/_extensions/_dwt.pyx", line 101, in pywt._extensions._dwt.idwt_single (pywt/_extensions/_dwt.c:5454)

  File "pywt/_extensions/_dwt.pyx", line 112, in pywt._extensions._dwt.idwt_single (pywt/_extensions/_dwt.c:4961)

ValueError: Coefficients arrays must have the same dtype.
```